### ISSUE TITLE
GitLens creates a `.gk` folder in the home dir.

### ISSUE THREAD

**Issue by lgarron** [gitkraken/vscode-gitlens#3660]
### Description

1. Use the extension like normal?

### GitLens Version

v15.6.0

### VS Code Version

Version: 1.94.0
Commit: d78a74bcdfad14d5d3b1b782f87255d802b57511
Date: 2024-10-02T13:08:12.626Z (6 days ago)
Electron: 30.5.1
ElectronBuildId: 10262041
Chromium: 124.0.6367.243
Node.js: 20.16.0
V8: 12.4.254.20-electron.0
OS: Darwin arm64 24.0.0

### Git Version

git version 2.45.0

### Logs, Screenshots, Screen Captures, etc

A `.gk` folder has started appearing in my home directory, containing a `repoMapping.json` file. This seems to be due to: https://github.com/gitkraken/vscode-gitlens/blob/b99d04597b46c0c59e9e56adb01db82a8974e089/src/env/node/pathMapping/sharedGKDataFolder.ts#L60

Creating a folder directly in the home dir is [not a great practice](https://github.com/b3nj5m1n/xdg-ninja). The appropriate thing would be to use the XDG conventions, possibly supplemented by platform-specific conventions.

VS Code may provide an API for this to extensions, else there are a bunch of libraries for this: https://www.npmjs.com/search?q=xdg

Alternatively, creating `repoMapping.json` inside e.g. `~/.data/gitlens` would be *much, much* better than the current behaviour.

**Comment by eamodio** [comment]
This is by-design as that folder and its configuration is shared between our (GitKraken) tools (GitLens, GitKraken Desktop, and GitKraken CLI)

**Comment by lgarron** [comment]
> This is by-design as that folder and its configuration is shared between our (GitKraken) tools (GitLens, GitKraken Desktop, and GitKraken CLI)

Does this mean you are not open to any changes to moving the default location for that folder? 😕

**Comment by eamodio** [comment]
Correct, our tools use that as a well-known location. It is also quite a common location for many tools.

**Comment by lgarron** [comment]
> Correct, our tools use that as a well-known location. It is also quite a common location for many tools.

This is disappointing to hear, but is there anything that would change your mind? I would be happy to send a PR that uses an updated location when the old location is not present.

Many projects already use XDG-compatible conventions, and many have transitioned after listening to their community. Some of us see openness to this kind of improvement as a litmus for how a project is willing to listen to feedback.

**Comment by eamodio** [comment]
Can you point me to other tools (vs code in particular) that do this? What would it look like? 

But as I said this isn't just a change that GitLens can make because our other tools rely on these same files.

**Comment by lgarron** [comment]
Unfortunately I don't have a large list of tools that follow the XDG conventions, as by definition I will not notice folders from them.

However, https://wiki.archlinux.org/title/XDG_Base_Directory lists a large number of projects that used to have this issue but many of which have either moved their default folder or added a configuration option to change the location.

Here is a selection based on a quick search of my GitHub notification history:

- https://github.com/chromium/trickuri/pull/18#event-2111051957
- https://github.com/pqrs-org/Karabiner-Elements/pull/114#event-908525492
- https://github.com/rubygems/rubygems/pull/3545#issuecomment-1283788635
- https://github.com/oven-sh/bun/pull/1351

**Comment by untainsYD** [comment]
> Correct, our tools use that as a well-known location. It is also quite a common location for many tools.

no it's not quite common, you are violating XDG DESKTOP PATH, you should google "common" folders before writing this nonsense 
@eamodio

**Comment by eamodio** [comment]
@untainsYD keep this constructive

**Comment by untainsYD** [comment]
> [@untainsYD](https://github.com/untainsYD) keep this constructive

oh, no, sorry, you are not constructive, you have broken the default path on Linux, and saying that it won't be fixed, coz you don't know how to set xdg data default path for Linux

@eamodio see https://github.com/microsoft/vscode/issues/162712, and `XDG_DATA_HOME` in https://wiki.archlinux.org/title/XDG_Base_Directory, https://specifications.freedesktop.org/basedir-spec/latest/

**Comment by BurnerWah** [comment]
> Can you point me to other tools (vs code in particular) that do this? What would it look like?
> 
> But as I said this isn't just a change that GitLens can make because our other tools rely on these same files.

I'm pretty sure that [SQLTools](https://github.com/mtxr/vscode-sqltools) follows the XDG spec, if you want VSCode-specific.

Also, fwiw [there's been a feedback suggestion for GitKraken itself to add XDG compliance](https://feedback.gitkraken.com/suggestions/197923/support-for-moving-the-config-directory-on-linux) since 2021.

I'd honestly be okay with even just having a `GITKRAKEN_HOME` environment variable so I can just put the `.gk` folder elsewhere. I don't actually use any of the other tools that rely on the `.gk` folder, so it's not really doing anything on my system. (well actually it doesn't exist because my home folder permissions prevent its creation entirely)

**Comment by untainsYD** [comment]
> well actually it doesn't exist because my home folder permissions prevent its creation entirely

Can you share how you achieved that?

**Comment by DusanLesan** [comment]
This is a deal breaker for me. If you do not start fixing it soon, I am out

**Comment by BurnerWah** [comment]
> > well actually it doesn't exist because my home folder permissions prevent its creation entirely
> 
> Can you share how you achieved that?

`chmod -w $HOME`
Only works on Linux, and it can break some desktop environments (I use GNOME on Fedora Linux, but last time I used KDE it wasn't happy), but if it works, it works very well.
As long as no mount-related things happen, everything either just silently errors out or crashes instead of creating dotfiles.

Combined with some wrapper scripts to give things that do create dotfiles an alternate home folder (I use a very overengineered solution to this but just changing what `$HOME` to something else for a moment works too), and a... probably excessive utilization of flatpaks, you'll end up with an immaculate home folder.

**Comment by jmtornetta** [comment]
Agreed, this is a bad convention. A lot of us version control the home folder or have scripts that run recursive actions on folders in $HOME on a cron/systemd. This creates more work.

**Comment by TilKenneth** [comment]
Hello from the Windows world. 
I have 19 dot-dirs and 8 dot-files, I also have these absolutely useless symlinks dangling in $HOME under threat Windows decides to make that its last stand.

Clutter.

I have to scroll to find my "Tools" directory, I even have %TOOLS% as a shortcut to skip this, the scrolling on my mouse has torn it old. Because .gitconfig, mercurial, the Bbash-4-Windows-dot-mess and their random configs everywhere, I somehow have .aws and some totally undecipherable .gk directory, with only hint is a json related to gitmapping, a need I presume, because another unrelated useless directory.

You have %AppData% and %LocalAppData%, you even have the whole Windows Registry to hide in aswell, you also have %ProgramData%, i mean there are dead symlinks morphing into another language depending who "looks" at them, dangling and unclicable, because of options.

I am not a daily linux driver, but I really, realllly do appreciate the XDG-env vars thingies, I can remap you - free to choose, free as in freedom to do what I want if it suits me.

Now that I came here, I do know I have this extension. Lets imagine I feel the freedom to remove it, I assume your product will clean up, properly, like most serious Windows uninstallers?

---

Anyway, I really just came here to say, stop this cluttering. Please have a nice week, and I will take my goodbye for now!

**Comment by brendanzab** [comment]
It would be great for it to be in a XDG-compliant location on macOS too (which is what Git itself does). That said, I’m already probably be moving to [Better Git Line Blame](https://marketplace.visualstudio.com/items?itemName=mk12.better-git-line-blame)… this is just the thing that’s pushed me to fully uninstall GitLens.

I’ve been working really hard to get all the dotfiles out of my top-level home directory, and it’s a bit frustrating to have another pop up, then have to track down where it’s coming from! It’s also particularly annoying in this case because `~/.gk` doesn’t give any sign that it’s coming from GitLens, so you’re left relying on search engines.

**Comment by amorphobia** [comment]
> I’m already probably be moving to Better Git Line Blame

Thank you! This is constructive!

**Comment by ebits21** [comment]
Just to chime in, I’m cleaning up my home folder today and wondering what the .gk and .gitkraken folders are that are junking up my home folder.

Please follow xdg standards. You can move your config I’m sure. Pleaseee.

**Comment by zackheil** [comment]
Sadly #3904 doesn't support XDG for macOS... we use that standard too.

**Comment by nzaytsev** [comment]
@zackheil - now the changes are in the pre-release version only and planned to be in 16.2

**Comment by zackheil** [comment]
> now the changes are in the pre-release version only and planned to be in 16.2

I'm sorry I guess I don't follow. What does that has to do with the changes made in that PR not allowing all *NIX operating systems to adhere to XDG standard?

**Comment by nzaytsev** [comment]
@zackheil - I'm sorry, I misunderstood your initial concern. I thought you expected the issue to be closed, tested it, and encountered incorrect behavior. Could you clarify your concerns, please?

The PR aims to support all systems; however, not all operating systems fully adhere to the XDG standard out of the box. As a result, we decided to use specific directories tailored to each OS as described in the documentation:

- [XDG standard](https://specifications.freedesktop.org/basedir-spec/latest/) is fully supported on Linux out of the box.
- For macOS, we followed the [official Apple documentation](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW28) and used the recommended folders.
- Windows does not natively support the XDG standard, so we followed common practices by using the LocalAppData directory

Could you please clarify why you think this might not work on macOS, or explain what you expect from this PR? Your feedback would be greatly appreciated.

**Comment by TilKenneth** [comment]
> [@zackheil](https://github.com/zackheil) - I'm sorry, I misunderstood your initial concern. I thought you expected the issue to be closed, tested it, and encountered incorrect behavior. Could you clarify your concerns, please?
> 
> The PR aims to support all systems; however, not all operating systems fully adhere to the XDG standard out of the box. As a result, we decided to use specific directories tailored to each OS as described in the documentation:
> 
> * [XDG standard](https://specifications.freedesktop.org/basedir-spec/latest/) is fully supported on Linux out of the box.
> * For macOS, we followed the [official Apple documentation](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW28) and used the recommended folders.
> * Windows does not natively support the XDG standard, so we followed common practices by using the LocalAppData directory
> 
> Could you please clarify why you think this might not work on macOS, or explain what you expect from this PR? Your feedback would be greatly appreciated.

Microsoft official regarding Known Folders https://learn.microsoft.com/en-us/windows/win32/shell/known-folders - comprehensive list; https://learn.microsoft.com/en-us/windows/win32/shell/knownfolderid 

see for descriptions https://learn.microsoft.com/en-us/windows/win32/shell/csidl

CSIDL_PROFILE
FOLDERID_Profile
Version 5.0. The user's profile folder. A typical path is C:\Users\username. Applications should not create files or folders at this level; they should put their data under the locations referred to by CSIDL_APPDATA or CSIDL_LOCAL_APPDATA. However, if you are creating a new Known Folder the profile root referred to by CSIDL_PROFILE is appropriate.

SIDL_APPDATA
FOLDERID_RoamingAppData
Version 4.71. The file system directory that serves as a common repository for application-specific data. A typical path is C:\Documents and Settings\username\Application Data.

CSIDL_LOCAL_APPDATA
FOLDERID_LocalAppData
Version 5.0. The file system directory that serves as a data repository for local (nonroaming) applications. A typical path is C:\Documents and Settings\username\Local Settings\Application Data.

**Windows Vista introduces new storage scenarios and a new user profile namespace. To address these new factors, the older system of referring to standard folders by a CSIDL value has been replaced. As of Windows Vista, those folders are referenced by a new set of GUID values called Known Folder IDs.** https://learn.microsoft.com/en-us/windows/win32/shell/knownfolderid 

https://learn.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetknownfolderpath
Remarks
This function replaces SHGetFolderPath. That older function is now simply a wrapper for SHGetKnownFolderPath.

**Comment by lgarron** [comment]
Just wanted to say thanks for shipping this, especially given the initial reluctance!

It's really nice to reduce the number of workarounds in my dotfiles, and it now makes me really *glad* every time I remember I don't have to worry about `~/.gk` anymore.
